### PR TITLE
Add cdktf `watch` command

### LINKED ISSUE #801: cdktf `watch` command
The `watch` command is supposed to watch source files and to automatically do a `cdktf deploy (--auto-approve)` when they change to support quickly iterating on infrastructure (especially useful for serverless use-cases).

### Community Note

* Please vote on this issue by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request
* Please do not leave "+1" or other comments that do not add relevant new information or questions, they generate extra noise for issue followers and do not help prioritize the request
* If you are interested in working on this issue or have submitted a pull request, please leave a comment

### Description

Add a `watch` command to the `cdktf-cli` package. When executed, this command is
long-running and keeps watching the source directory for changes to files. When
a change is detected, it will run the `app` script of the `cdktf.json` and to
build and synthesize the application.  
After running the `synth` it will compare the `cdk.tf.json` output with a
previous version (probably using hashes kept
in-memory)
and determine whether changes need to be deployed. If there are any, it will
automatically deploy them (auto-approving), ideally also skipping the
`diff`/`plan` phase to save time.  
To make sure the user knows that the initial iteration of the `watch` command
deploys automatically, the `--auto-approve` flag is required on the `watch`
command until it supports a manual approval step before deploying. So, in the
first iteration only `cdktf watch --auto-approve` will work but `cdktf watch`
will fail with a short note about this behaviour.  
By using supplying `-refresh=false` to Terraform on subsequent deploys, we can
possibly further speed up things.

#### Determining the source directory

For a first iteration the command could watch any file in the same directory as
the `cdktf.json` file, except for the build / synth output of the CDKTF and
except for files specified in the `.gitignore` file.

If this might impose problems, the source directory could also be inferred for
the target language, however it might be useful to expose an option (on the CLI
command and/or in the `cdktf.json` config) as the code structure used by users
could differ quite substantially.

#### Handling subsequent changes

If new file updates are detected while a deploy is currently running a second
deploy should be queued so that changes can always be expected to get applied
(even if they take longer this way). In future iterations this could be made
smarter (e.g. just redoing the `synth` step if the deploy itself wasn't running
yet).

#### Handling errors

If an error happens, the `watch` should continue running but display a message
to the user that informs him that something failed. Possible errors include:

- resources can't be applied properly for some reason
- state is locked and stuck
- cloud credentials are expired

### Todos for the first iteration
- [ ] Add new CLI command that achieves `cdktf watch --auto-approve`
- [ ] `watch` command takes a stack as argument or uses the only one available if there's just one
- [ ] Watches files not in `.gitignore`
- [ ] Compiles application and runs `synth` when source files changed
- [ ] Hashes the `cdk.tf.json` for a stack and keeps the hash in memory
- [ ] Does not hash the `stacktrace` part of the `cdk.tf.json` which does not affect the deploy *or* follow-up issue created for this improvement
- [ ] Deploys using a locally running Terraform CLI
- [ ] Skips the plan phase when deploying with `--auto-approve` to speed up things
- [ ] Supports subsequent code changes while a deploy is running
- [ ] Follow-up issue created for supporting remote execution in Terraform Cloud?
- [ ] Handle error case: apply failed (e.g. HTTP 400 / invalid resource configuration)
- [ ] Handle error case: Terraform state is locked and stuck
- [ ] Handle error case: Cloud credentials are expired (of a Terraform provider)
- [ ] Add tests for watch command
- [ ] Add docs for using the `watch` command
- [ ] Add follow-up issues for improvements listed below
- [ ] … _probably more_ …

### Improvements for future iterations

The following sections contain improvements for the aforementioned first minimal
viable implementation. They will probably be converted into follow-up issues after the first iteration of this feature is finished (depending on whether they are still relevant then).

#### Manual Approval

When invoking `cdktf watch` without the `--auto-approve` flag the user has to
confirm applying the changes by pressing the enter key.

#### Detecting which resources changed

When checking the new `cdk.tf.json` against the previous version, we could check
which resources were changed and supply them as a
[-target](https://www.terraform.io/docs/cli/commands/plan.html#resource-targeting)
to the Terraform CLI. As we don't expect outside changes to our resources while
using `watch` we could save the time Terraform would take to refresh all the
other resources. This would further speed up the deployment of the actual
changes and improve the user experience.  
This might possibly not be worth the effort if supplying `-refresh=false` to
subsequent runs already speeds up things enough. However we might still want to
test this.

#### Make use of target language tooling for recompiling

Most of the target languages of the CDKTF have tooling for watching the source
files and incrementally compiling the resulting binary (e.g. `tsc --watch`,
`gradle --watch-fs` [Maven does not seem to support watching], `gomon`). We
could make use of this to speed up the build and synth part of the `cdktf watch`
command. We could e.g. add a `watch` option to the `cdktf.json` config which
contains a command that is spawned by the `cdktf watch` command.  
How exactly the `watch` command would then do the `synth` has still to be
explored.

### PR DESCRIPTION
**Pull request by ansgarm**
resolves #801 

~ work in progress ~

### CODE REVIEW

File: packages/cdktf-cli/bin/cmds/helper/synth-stack.ts [hunk 1]
```diff
@@ -26,7 +26,8 @@ interface ManifestJson {
 export class SynthStack {
   public static async synth(
     command: string,
-    outdir: string
+    outdir: string,
+    graceful = false // will not exit the process but rethrow the error instead
```
Comment (DanielMSchmidt) [inline packages/cdktf-cli/bin/cmds/helper/synth-stack.ts:30]: Nitpick: We need graceful due to the `process.exit` call as far as I understand. We want to get rid of the process.exit midterm since they interfere with telemetry and possibly other things. Why not switch to throwing errors here in general?

File: packages/cdktf-cli/bin/cmds/ui/watch.tsx [hunk 2]
```diff
@@ -0,0 +1,271 @@
...
+  error?: GraphQLWatchState["error"];
+}) => {
+  const [didDeployAlready, setDidDeployAlready] = useState(false);
+  useEffect(() => {
+    if (status === "DEPLOYING" && !didDeployAlready) setDidDeployAlready(true);
+  }, [didDeployAlready, status]);
+
+  let statusText = <Text>{status}</Text>;
+  const stack = stackName ? <Text color="blueBright">{stackName}</Text> : null;
+  switch (status) {
```
Comment (DanielMSchmidt) [inline packages/cdktf-cli/bin/cmds/ui/watch.tsx:27]: Nitpick: moving this into a function would make the component more readable IMHO :)

File: packages/cdktf-cli/bin/cmds/ui/watch.tsx [hunk 3]
```diff
@@ -0,0 +1,271 @@
...
+    throw new Error(`${error.origin}: ${error.message}`);
+  }
+  return (
+    <Text color="redBright">
+      {error.origin}: {stripAnsi(error.message)}
+    </Text>
+  );
+};
+
+interface WatchConfig {
```
Comment (DanielMSchmidt) [inline packages/cdktf-cli/bin/cmds/ui/watch.tsx:192]: Nitpick: https://www.graphql-code-generator.com/ could help here :)

File: packages/cdktf-cli/bin/cmds/watch.ts [hunk 4]
```diff
@@ -0,0 +1,74 @@
...
+  public async handler(argv: any) {
+    await displayVersionMessage();
+    const command = argv.app;
+    const outdir = argv.output;
+    const autoApprove = argv.autoApprove;
+    const stack = argv.stack;
+
+    if (!autoApprove) {
+      console.error(
+        chalkColour`{redBright ERROR: The watch command currently always automatically deploys and approves changes. Please add the --auto-approve flag to confirm you understand this behaviour.}`
```
Comment (DanielMSchmidt) [inline packages/cdktf-cli/bin/cmds/watch.ts:54]: This reads a bit condescending to me 🤔 

```suggestion
        chalkColour`{redBright ERROR: The watch command currently always automatically deploys and approves changes. If this work with your workflow, please add the --auto-approve flag to confirm.}`
```

File: packages/cdktf-cli/bin/cmds/watch.ts [hunk 5]
```diff
@@ -0,0 +1,74 @@
...
+      );
+      process.exit(1);
```
Comment (DanielMSchmidt) [inline packages/cdktf-cli/bin/cmds/watch.ts:56]: Nitpick: We could also throw / early return here right?

File: packages/cdktf-cli/lib/client/react.tsx [hunk 6]
```diff
@@ -0,0 +1,97 @@
...
+import { getMainDefinition } from "@apollo/client/utilities";
+import { WebSocketLink } from "@apollo/client/link/ws";
+import { onError } from "@apollo/client/link/error";
+import { logger } from "../logging";
+import { Text } from "ink";
+import { startServer } from "../server";
+import React, { ReactNode, useEffect, useState } from "react";
+import { SubscriptionClient } from "subscriptions-transport-ws";
+
+const useAsyncEffect = (fun: () => Promise<void>, deps: Array<any>) => {
```
Comment (DanielMSchmidt) [inline packages/cdktf-cli/lib/client/react.tsx:21]: Why is this here, I don't understand the need for this function :)

File: packages/cdktf-cli/lib/server/WatchClient.ts [hunk 7]
```diff
@@ -0,0 +1,353 @@
...
+        // return terraformCloud;
+      }
+    }
+    return new TerraformCli(stack);
+  }
+
+  public async start() {
+    this.running = true;
+
+    const gitignored = await readGitignore(process.cwd());
```
Comment (DanielMSchmidt) [inline packages/cdktf-cli/lib/server/WatchClient.ts:260]: Nitpick: There is no cheap way to listen to gitignores as well and reevaluate this, right? I'm always annoyed when I have to restart sth due to such a config change

File: packages/cdktf-cli/lib/server/server.ts [hunk 8]
```diff
@@ -0,0 +1,136 @@
...
+
+          if (watchClient && (await watchClient.isRunning())) {
+            console.warn(
+              "There is already a watch running, you will receive the updates from that run but no new one is started."
+            );
+          } else {
+            try {
+              watchClient = new WatchClient(args.inputs);
+
+              // TODO: subscribe returns an unsubscribe function but that connect reliably
```
Comment (DanielMSchmidt) [inline packages/cdktf-cli/lib/server/server.ts:54]: I show how to use RxJS instead in this video, but that might be overkill: https://www.youtube.com/watch?v=7SOwFE9lU-g

**Review by DanielMSchmidt — COMMENTED** [review]
That's a really cool new feature 🙌

File: packages/cdktf-cli/bin/cmds/helper/synth-stack.ts (hunk 1 above)
Comment (ansgarm) [reply packages/cdktf-cli/bin/cmds/helper/synth-stack.ts:30]: Because I did not know what kind of repercussion that might cause for the existing commands using it 😅

File: packages/cdktf-cli/bin/cmds/ui/watch.tsx (hunk 2 above)
Comment (ansgarm) [reply packages/cdktf-cli/bin/cmds/ui/watch.tsx:27]: You mean multiple different components instead of this big switch?

File: packages/cdktf-cli/bin/cmds/ui/watch.tsx (hunk 3 above)
Comment (ansgarm) [reply packages/cdktf-cli/bin/cmds/ui/watch.tsx:192]: Jup, that's something for the future – I mentioned generating the types somewhere else in some Todo comment.

File: packages/cdktf-cli/bin/cmds/watch.ts (hunk 4 above)
Comment (ansgarm) [reply packages/cdktf-cli/bin/cmds/watch.ts:54]: You're right – will adjust this.

Comment (ansgarm) [reply packages/cdktf-cli/bin/cmds/watch.ts]: Changed it to
> To make this behaviour explicit the --auto-approve flag must be set
What do you think?

File: packages/cdktf-cli/lib/client/react.tsx (hunk 6 above)
Comment (ansgarm) [reply packages/cdktf-cli/lib/client/react.tsx:21]: Only to save myself a level of indentation because I cannot directly pass an async function to useEffect 😅

File: packages/cdktf-cli/bin/cmds/watch.ts (hunk 5 above)
Comment (ansgarm) [reply packages/cdktf-cli/bin/cmds/watch.ts:56]: Yes, I kinda copied that behaviour from other commands I guess 😅

File: packages/cdktf-cli/lib/server/WatchClient.ts (hunk 7 above)
Comment (ansgarm) [reply packages/cdktf-cli/lib/server/WatchClient.ts:260]: Comes with complexity that I just wanted to avoid for the first version. Another thing that we currently also don't handle are nested `.gitignore` files.

File: packages/cdktf-cli/lib/server/server.ts (hunk 8 above)
Comment (ansgarm) [reply packages/cdktf-cli/lib/server/server.ts:54]: Hehe, neat idea – but working with Observables would be another concept that we'd introduce in our codebase then 😄 But maybe that's a neat way of doing that at some point – who know 😁

Comment (DanielMSchmidt) [reply packages/cdktf-cli/lib/server/server.ts:54]: :D yeah not really worth it here. But it also explains how to implement your own async iterable which could be useful here 🤔

File: packages/cdktf-cli/lib/client/react.tsx (hunk 6 above)
Comment (DanielMSchmidt) [reply packages/cdktf-cli/lib/client/react.tsx:21]: ah ok

File: packages/cdktf-cli/lib/server/server.ts (hunk 8 above)
Comment (ansgarm) [reply packages/cdktf-cli/lib/server/server.ts:54]: Ah, nice! Gonna resolve this for now anyway. bye 👋😄

Comment (ansgarm) [reply packages/cdktf-cli/lib/server/server.ts:54]: Oh, it was resolved already

File: packages/cdktf-cli/bin/cmds/watch.ts (hunk 4 above)
Comment (DanielMSchmidt) [reply packages/cdktf-cli/bin/cmds/watch.ts]: Way better!

File: packages/cdktf-cli/bin/cmds/watch.ts (hunk 5 above)
Comment (ansgarm) [reply packages/cdktf-cli/bin/cmds/watch.ts:56]: Going to ignore this in favor of a uniform solution addressing all out exit's in the future 😬

File: packages/cdktf-cli/bin/cmds/ui/watch.tsx (hunk 2 above)
Comment (ansgarm) [reply packages/cdktf-cli/bin/cmds/ui/watch.tsx:27]: Going to keep it this way for now.

**Review by DanielMSchmidt — COMMENTED** [review]
- When I run this I see starting server and connecting messages. If I didn't read through the implementation I'd be confused to what server is starting and what it's connecting to. 
- I see errors twice currently
Screenshot 2021-07-29 at 13 57 33
- UI wise these should be in separate lines I beli
Screenshot 2021-07-29 at 13 59 04
eve
- It does not show the last elements being deployed 
Screenshot 2021-07-29 at 14 05 14
- Terraform assets seem to be not included in the watch, I changed sth in my application I deploy via NullResource and include via TerraformAsset but no update was triggered

File: docs/working-with-cdk-for-terraform/watch.md [hunk 9]
```diff
@@ -0,0 +1,31 @@
...
+
+> ⚠️ Warning: The `watch` command is a new experimental command. **ONLY use in development environments**
+
+The `watch` command watches a directory for changes and automatically synthesizes and deploys changes as they happen. It allows for rapid iterations when developing infrastructure, especially when working with serverless services. It currently supports only one stack at a time.
+
+## Using `watch`
+
+Watch reads your root `.gitignore` file to determine which files trigger a `synth` and it uses the `outdir` of your `cdktf.json` (or the default `cdktf.out`) to do so for a deploy. It will generate a checksum of the subdirectory containing the Terraform code for your stack to skip deploys if the synthesized Terraform config did not change.
+
+### Checking your root `.gitignore`
```
[the remark this answers is not in the document]
Comment (ansgarm) [reply docs/working-with-cdk-for-terraform/watch.md:11]: There's a todo to create a follow up issue in the checklist over here: https://github.com/hashicorp/terraform-cdk/issues/801
So I'm going to resolve this :)

File: packages/cdktf-cli/bin/cmds/helper/synth-stack.ts [hunk 10]
```diff
@@ -22,10 +22,14 @@ interface ManifestJson {
   stacks: StackManifest[];
 }
 
+type SynthOrigin = "watch";
```
[the remark this answers is not in the document]
Comment (ansgarm) [reply packages/cdktf-cli/bin/cmds/helper/synth-stack.ts:25]: huh? That's just some type to narrow down what could be expected in the `synthOrigin` payload for telemetry.

File: packages/cdktf-cli/bin/cmds/ui/watch.tsx [hunk 11]
```diff
@@ -0,0 +1,271 @@
...
+    case "INITIALIZING":
+      statusText = <Text>Initializing Terraform for {stack || "stack"}</Text>;
+      break;
+    case "CONNECTING":
+      statusText = <Text>Connecting</Text>;
+      break;
+  }
+  return (
+    <Box>
+      {status === "IDLE" ? (
```
[the remark this answers is not in the document]
Comment (ansgarm) [reply packages/cdktf-cli/bin/cmds/ui/watch.tsx:52]: Idle is a valid status to happen at anytime.
I needed to create an empty Text to be able to remove and readd the spinner – else it wouldn't re-appear 😅

File: packages/cdktf-cli/bin/cmds/ui/watch.tsx [hunk 12]
```diff
@@ -0,0 +1,271 @@
...
+      aprv: autoApprove,
+    },
+  });
+  const watchedStackName =
+    targetStack ||
+    data?.watch.stacks.find((_: any, idx: number) => idx === 0)?.name;
+  const status = data?.watch.status || "CONNECTING";
+
+  if (error) console.error("uncaught error ", error);
+  // TODO: display error if an error happened
```
[the remark this answers is not in the document]
Comment (ansgarm) [reply packages/cdktf-cli/bin/cmds/ui/watch.tsx:253]: Well, the console.error() sorta does that 😅

File: packages/cdktf-cli/bin/cmds/watch.ts [hunk 13]
```diff
@@ -0,0 +1,74 @@
...
+        desc: "Deploy stack which matches the given id only. Required when more than one stack is present in the app",
+        type: "string",
+      })
+      .option("app", {
+        default: config.app,
+        required: true,
+        desc: "Command to use in order to execute cdktf app",
+        alias: "a",
+      })
+      .option("output", {
```
[the remark this answers is not in the document]
Comment (ansgarm) [reply packages/cdktf-cli/bin/cmds/watch.ts:31]: The same as for `diff` and `deploy` – I think it allows overriding the output directory where synthed stacks end up

File: packages/cdktf-cli/bin/cmds/watch.ts [hunk 14]
```diff
@@ -0,0 +1,74 @@
...
+  public async handler(argv: any) {
+    await displayVersionMessage();
+    const command = argv.app;
+    const outdir = argv.output;
+    const autoApprove = argv.autoApprove;
+    const stack = argv.stack;
+
+    if (!autoApprove) {
+      console.error(
+        chalkColour`{redBright ERROR: The watch command currently always automatically deploys and approves changes. To make this behaviour explicit the --auto-approve flag must be set}`
```
[the remark this answers is not in the document]
Comment (ansgarm) [reply packages/cdktf-cli/bin/cmds/watch.ts:54]: Because UsageError did not exist yet, when I wrote this line :D

File: packages/cdktf-cli/lib/server/WatchClient.ts [hunk 15]
```diff
@@ -0,0 +1,395 @@
...
+    await terraform.init();
+    this.needsInit = false;
+  }
+
+  private async runDeploy() {
+    const { workingDirectory, name } = await this.getTargetStack();
+    console.log({ workingDirectory });
+    const newTargetStackHash = hashPath(workingDirectory);
+    if (this.lastTargetStackHash === newTargetStackHash) {
+      console.log(
```
[the remark this answers is not in the document]
Comment (ansgarm) [reply packages/cdktf-cli/lib/server/WatchClient.ts:156]: All logs inside the server child process are logged through logger via the code that spawned it. So everything in server.ts and WatchClient.ts just uses console :)

File: docs/working-with-cdk-for-terraform/watch.md [hunk 16]
```diff
@@ -0,0 +1,31 @@
...
+
+Your root `.gitignore` needs to contain all files that are generated by a synth and a deploy. If that is not the case the watch will trigger itself and continuously try to deploy. Although Git supports nesting `.gitignore` files, the watch command currently only reads the root `.gitignore` file.
+
+### Checking your environment
+
+Watch should only be used for development environments. It is best to make sure that the terminal you're running watch in has no access keys that allow the cdktf-cli to deploy to your production environment.
```
[the remark this answers is not in the document]
Comment (ansgarm) [reply docs/working-with-cdk-for-terraform/watch.md:17]: Because if you e.g. accidentally rename your stack you could affect your production environment.
Yes, it is your choice, but you "should not" 😁

**Comment by ansgarm** [comment]
> I see errors twice currently

@DanielMSchmidt Have you set the CDKTF_LOG_LEVEL to something other than the default?

File: packages/cdktf-cli/lib/server/WatchClient.ts [hunk 17]
```diff
@@ -0,0 +1,395 @@
...
+  private async runInit() {
+    this.updateState({ status: "INITIALIZING" });
+    const terraform = await this.getTerraform();
+    await terraform.init();
+    this.needsInit = false;
+  }
+
+  private async runDeploy() {
+    const { workingDirectory, name } = await this.getTargetStack();
+    console.log({ workingDirectory });
```
[the remark this answers is not in the document]
Comment (ansgarm) [reply packages/cdktf-cli/lib/server/WatchClient.ts:153]: It is on purpose, but I could reword it to make better debug logs. As noted in another reply: `console.log()` is used for all logging of the server child process.

File: packages/cdktf-cli/lib/server/WatchClient.ts [hunk 18]
```diff
@@ -0,0 +1,395 @@
...
+          ["-refresh=false"]
+        );
+      }
+      // deployment was successful -> update hash
+      this.lastTargetStackHash = newTargetStackHash;
+    } catch (e) {
+      // When new providers where added or the stack name might've
+      // been changed (when using just a single stack) we need to
+      // re-init Terraform. We'll do this proactively on any error.
+      this.needsInit = true;
```
[the remark this answers is not in the document]
Comment (ansgarm) [reply packages/cdktf-cli/lib/server/WatchClient.ts:196]: Because currently after a recoverable error the WatchClient will be IDLE until a synth or a deploy is triggered. It will then run an init after the synth but before another deploy. So in the case of synth the status would be SYNTHESIZING and then DEPLOYING while needsInit needs to be retained until the synth finished.
But yeah, I'd love to refactor this into a proper state machine at some point. I also think this is ugly.

File: packages/cdktf-cli/lib/server/WatchClient.ts [hunk 19]
```diff
@@ -0,0 +1,395 @@
...
+
+  private async getTargetStack(): Promise<Stack> {
+    const { stacks } = this.state;
+    if (stacks.length === 0) {
+      throw new RecoverableError(
+        "Cannot determine target stack when there are no stacks",
+        this.state.status
+      );
+    }
+    let stack: Stack | undefined = stacks[0];
```
[the remark this answers is not in the document]
Comment (ansgarm) [reply packages/cdktf-cli/lib/server/WatchClient.ts:225]: Good idea 👍

File: packages/cdktf-cli/lib/server/WatchClient.ts [hunk 20]
```diff
@@ -0,0 +1,395 @@
...
+
+  public async queueAction(action: Action, priority = false) {
+    // only enqueue if not already enqueued anyway
+    if (!this.actionQueue.includes(action)) {
+      // priority = run next
+      if (priority) this.actionQueue.unshift(action);
+      else this.actionQueue.push(action);
+    }
+
+    console.log(`currently queued actions: ${this.actionQueue.join(", ")}`);
```
[the remark this answers is not in the document]
Comment (ansgarm) [reply packages/cdktf-cli/lib/server/WatchClient.ts:335]: no, this is part of the debug logging of the server

File: packages/cdktf-cli/lib/server/WatchClient.ts [hunk 21]
```diff
@@ -0,0 +1,395 @@
...
+    ["change", "add"].forEach((event) => {
+      if (this.sourceFileWatcher)
+        this.sourceFileWatcher.on(event, (path) => {
+          console.log(`synth - path changed: ${event} ${path}`);
+          this.queueAction("SYNTH");
+        });
+    });
+
+    // If out dir files change we queue a deploy to be run
+    // TODO: only watch stack.workingDirectory (as soon as stack is known)
```
[the remark this answers is not in the document]
Comment (ansgarm) [reply packages/cdktf-cli/lib/server/WatchClient.ts:304]: it was obsolete, reworded it.

File: packages/cdktf-cli/lib/server/WatchClient.ts [hunk 22]
```diff
@@ -0,0 +1,395 @@
...
+      else this.actionQueue.push(action);
+    }
+
+    console.log(`currently queued actions: ${this.actionQueue.join(", ")}`);
+  }
+
+  private async startHandlingActions() {
+    while (this.isRunning()) {
+      const nextAction: Action | undefined = this.actionQueue.shift();
+      if (nextAction) console.log({ nextAction });
```
[the remark this answers is not in the document]
Comment (ansgarm) [reply packages/cdktf-cli/lib/server/WatchClient.ts:341]: that's for server debugging – reworded it.

File: packages/cdktf-cli/lib/server/WatchClient.ts [hunk 23]
```diff
@@ -0,0 +1,395 @@
...
+type WatchStateStatus =
+  | "IDLE"
+  | "SYNTHESIZING"
+  | "INITIALIZING"
+  | "PLANNING"
+  | "DEPLOYING";
+
+export type WatchErrorOrigin = WatchStateStatus;
+
+class RecoverableError extends Error {
```
[the remark this answers is not in the document]
Comment (ansgarm) [reply packages/cdktf-cli/lib/server/WatchClient.ts:47]: At some point we might want to, but for now the concept of RecoverableErrors makes only sense to me in the context of Watch. These are errors that would normally stop a deploy but in Watch will just cause it to wait for new file changes – so until the user fixes that error.

**Comment by DanielMSchmidt** [comment]
@ansgarm I don't think so, I believe it was unset. But shouldn't the logs go only into the cdktf.log file anyways?

File: packages/cdktf-cli/bin/cmds/ui/watch.tsx (hunk 11 above)
[the remark this answers is not in the document]
Comment (ansgarm) [reply packages/cdktf-cli/bin/cmds/ui/watch.tsx:52]: Ah, I misunderstood "early return" – changed it, nice idea 🙌

File: packages/cdktf-cli/bin/cmds/ui/watch.tsx (hunk 12 above)
[the remark this answers is not in the document]
Comment (ansgarm) [reply packages/cdktf-cli/bin/cmds/ui/watch.tsx:253]: removed todo

File: packages/cdktf-cli/lib/server/server.ts [hunk 24]
```diff
@@ -0,0 +1,136 @@
...
+  autoApprove: boolean;
+  targetStack?: string;
+};
+
+const pubsub = new PubSub();
+const WATCH_EVENT = "WATCH";
+
+const PORT = process.argv[2];
+
+console.log("server started with cwd:", process.cwd());
```
[the remark this answers is not in the document]
Comment (ansgarm) [reply packages/cdktf-cli/lib/server/server.ts:26]: yes, this was in fact left

File: packages/cdktf-cli/lib/server/server.ts [hunk 25]
```diff
@@ -0,0 +1,136 @@
...
+          // while startup reach the client
+          const iterator = pubsub.asyncIterator([WATCH_EVENT]);
+
+          if (watchClient && watchClient.isRunning()) {
+            console.warn(
+              "There is already a watch running, you will receive the updates from that run but no new one is started."
+            );
+          } else {
+            try {
+              watchClient = new WatchClient(args.inputs);
```
[the remark this answers is not in the document]
Comment (ansgarm) [reply packages/cdktf-cli/lib/server/server.ts:52]: I will create a follow up issue

File: packages/cdktf-cli/bin/cmds/watch.ts (hunk 14 above)
[the remark this answers is not in the document]
Comment (ansgarm) [reply packages/cdktf-cli/bin/cmds/watch.ts:54]: Ditched the "currently"

**Comment by ansgarm** [comment]
> @ansgarm I don't think so, I believe it was unset. But shouldn't the logs go only into the cdktf.log file anyways?

@DanielMSchmidt If you set the log level but not the DISABLE_LOGGING flag, it will print the errors afaik

**Comment by DanielMSchmidt** [comment]
> @DanielMSchmidt If you set the log level but not the DISABLE_LOGGING flag, it will print the errors afaik

Yeah that could have been the case :)

**Comment by ansgarm** [comment]
@DanielMSchmidt 
Fixed multiple resources that were displayed side by side ✅
flex_direction

**Comment by ansgarm** [comment]
Also massive spoiler for the Office Hours later on 😉

**Review by DanielMSchmidt — APPROVED** [review]

**Comment by ansgarm** [comment]
> Terraform assets seem to be not included in the watch, I changed sth in my application I deploy via NullResource and include via TerraformAsset but no update was triggered

@DanielMSchmidt Is that file you changed by chance in your `.gitignore`?

Theoretically it should watch every file that is not in your `.gitignore` for synthesizing and it should watch all files in the working directory of your target stack for deploying.

**Comment by DanielMSchmidt** [comment]
> @DanielMSchmidt Is that file you changed by chance in your .gitignore?

No I referenced the docker backend with TerraformAsset that is in ../../application/backend and changed the index.js in there (in the docker e2e example)

**Comment by ansgarm** [comment]
Ah, alright, that's why. It will only watch files that are located inside the CDKTF working directory / project directory but not anywhere outside 😅
However, for Assets we could make this work because we could use the synth process to know which asset source directories exist (if that would e.g. be included in a future version of the manifest file)
